Allow app check code for deprecated classes

nickvergessen opened the pull request:
Old behaviour (without parameter):

```
nickv@nickv-think ~/ownCloud/master/core $ ./occ app:check-code activity
App is compliant - awesome job!
```

With deprecated parameter:

```
nickv@nickv-think ~/ownCloud/master/core $ ./occ app:check-code activity --deprecated
Analysing /home/nickv/ownCloud/master/repos/activity/templates/stream.app.navigation.php
 1 errors
    line   14: == - is discouraged
Analysing /home/nickv/ownCloud/master/repos/activity/controller/settings.php
 2 errors
    line   31: OCP\IConfig - deprecated class must not be imported with a use statement
    line  112: == - is discouraged
Analysing /home/nickv/ownCloud/master/repos/activity/lib/data.php
 1 errors
    line   27: OCP\DB - deprecated class must not be imported with a use statement

Analysing /home/nickv/ownCloud/master/repos/activity/appinfo/update.php
 3 errors
    line   14: OCP\DB - Static method of deprecated class must not be called
    line   24: OCP\DB - Static method of deprecated class must not be called
    line   26: OCP\DB - Static method of deprecated class must not be called
App uses deprecated functionality
```

Requested at https://mailman.owncloud.org/pipermail/devel/2015-June/001301.html

I think it's quite useful.
Also fixing a bug (use statements could circumvent blacklisted classes)

@DeepDiver1975 @MorrisJobke @LukasReschke 
### Todo
- [x] Add support for deprecated classes
- [x] Add support for deprecated constants
- [x] Add support for deprecated functions
- [x] Add support for deprecated methods

nickvergessen committed:
Allow app:check-code to check for deprecated methods

nickvergessen committed:
Correctly handle use statements

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/c39cc05c-c32b-40ac-8ef8-0930b65bdc52) was created.

nickvergessen committed:
Add deprecation version to the list

nickvergessen committed:
Add support for deprecated constants

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/fde3a87a-93dd-4960-b406-64a5bf519317) was created.

nickvergessen committed:
Allow checking for functions

nickvergessen committed:
Check for methods as good as possible

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/16394c7c-ce39-4cb9-9634-ceee435e4e90) was created.

rullzer commented:
Awesome work.
We should also try to get a checker for out internal code. To make sure non deprecated functions/methods do not call deprecated functions/methods..

PVince81 commented:
Needs rebase

DeepDiver1975 commented:
@nickvergessen please rebase an let's move this in - THX

nickvergessen commented:
Btw, does anyone know something that got deprecated in the last 21days?

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/e4551bb7-a428-4425-a97b-11e74bd52b4d) was created.

DeepDiver1975 commented:
> Btw, does anyone know something that got deprecated in the last 21days?

Not that I'm aware of

nickvergessen commented:
Rebased and added the open task as todo.

Should be okay to merge, it doesn't catch everything, but at least static methods are done.

nickvergessen commented on core/command/app/checkcode.php:
```diff
@@ -72,6 +85,8 @@ protected function execute(InputInterface $input, OutputInterface $output) {
 		if (empty($errors)) {
 			$output->writeln('<info>App is compliant - awesome job!</info>');
+		} elseif ($input->getOption('deprecated')) {
+			$output->writeln('<comment>App uses deprecated functionality</comment>');
```
missing non zero-exit code, hmpf

Xenopathic commented on core/command/app/checkcode.php:
```diff
@@ -37,12 +40,22 @@ protected function configure() {
 		$appId = $input->getArgument('app-id');
-		$codeChecker = new \OC\App\CodeChecker();
+		if ($input->getOption('deprecated')) {
+			$codeChecker = new DeprecationCodeChecker();
```
I would highly recommend use of the Wrapper pattern here, instead of subclassing CodeChecker with the desired functionality. It means that if we introduce some other functionality at a later point, we just re-wrap the CodeChecker instead of needing 6 (!!!) subclasses. It quickly spirals out of control after that...

nickvergessen replied:
/me wraps Xenopathic into a decorator....

nickvergessen replied:
*done

nickvergessen committed:
Use non-zero exit code

nickvergessen committed:
Restructor the code into different classes instead of extending

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/3dcbc829-4b65-4407-9adc-f64e6529e8cb) was created.

Xenopathic commented on core/command/app/checkcode.php:
```diff
@@ -37,12 +41,24 @@ protected function configure() {
+		if ($input->getOption('deprecated')) {
+			$list = new DeprecationList();
+		} else {
+			$list = new PrivateList();
```
:see_no_evil: This isn't actually the Decorator pattern... What happens if we introduce some more checking, that we might need to combine with DeprecationList and/or PrivateList? You have the same problem, needing many classes and subclasses to cover all possibilities.

Have a BasicList class, whose methods return empty arrays, then let DeprecationList and PrivateList take an ICheckList in the constructor, and forward the methods to that wrapped ICheckList. Then you create a list at the start (`$list = new BasicList()`), then if we want to test for deprecation replace the object with the wrapped DeprecationList (`$list = new DeprecationList($list)`), etc for all features.

Sorry to be such a pain, but I can see this needing a refactor in the (near?) future if it isn't done now.

nickvergessen replied:
Right, mixed it up a bit. Should be fixed now

nickvergessen committed:
Use the decorator pattern

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/605382ee-788b-4356-89cc-5a81e85a6623) was created.

Xenopathic commented on core/command/app/checkcode.php:
```diff
@@ -37,12 +43,26 @@ protected function configure() {
+				'deprecated',
+				'd',
+				InputOption::VALUE_NONE,
+				'check the specified app'
```
Copypasta? :laughing:

Xenopathic commented:
Nice! Any reason why it's DeprecatedCheck **or** PrivateCheck? Can we not have PrivateCheck enabled by default (controllable by a new parameter would be cool) and enable DeprecatedCheck as well if the option is set? Or was the intention to do this?

I can see `./occ app:check-code --deprecated --no-private --check-foobar` being really powerful when the app checker gets more features...

rullzer commented:
I agree with @Xenopathic lets enable the extra checks by default.
Programmers are lazy so better enable as much by default as possible.

nickvergessen commented:
My idea was to be able to fail travis on private class usage and make travis "allow failure" for deprecated usage... so as long as it is possible to do those two separatly I'm fine with it

Xenopathic commented:
@nickvergessen Can I push to this branch? I'll implement the selectable features thing I mentioned above if you want :smile:

MorrisJobke commented:
@Xenopathic Do it. @nickvergessen is usually on vacation right now :P

Commit:
Allow specifying the checkers used in app:check-code

By default, all are enabled, but specific ones can be selected through command
line options.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/f2cfce49-b4f9-4f04-be3b-5433570bfc4e) was created.

Xenopathic commented:
You can now specify the checkers you want enabled, with `--checker=...` or `-c...`. So to only enable private class and strong comparison checks, use `./occ app:check-code -cprivate -cstrong-comparison`.

MorrisJobke commented:
```
Analysing /home/mjob/Projekte/owncloud/master/apps2/music/app/music.php
 1 errors
    line  236: OCP\User::getUser - Method of deprecated private class must not be called
Analysing /home/mjob/Projekte/owncloud/master/apps2/music/middleware/ampachemiddleware.php
 1 errors
    line   71: OC_Util - Static method of deprecated private class must not be called
```

* :x: `OCP\User` is not a private class, but is correctly deprecated.
* :white_check_mark: `OC_Util`private and deprecated

Xenopathic commented:
@MorrisJobke Ah, that's a side effect of the description concatenation. If both `DeprecationCheck` and `PrivateCheck` are enabled, any failures will be marked as `deprecated private class`, regardless of which checker actually found the error.

MorrisJobke commented:
@Xenopathic Can we fix this?

Xenopathic commented:
@MorrisJobke I have an idea...

nickvergessen commented on core/command/app/checkcode.php:
```diff
@@ -37,12 +47,30 @@ protected function configure() {
+				'checker',
+				'c',
+				InputOption::VALUE_REQUIRED | InputOption::VALUE_IS_ARRAY,
+				'enable the specified checker',
```
Because this is what people see as help, we should maybe add the available checkers here?

Xenopathic replied:
The help text displays the default value, which is currently all the available checkers (all are enabled by default). If we include some non-default-on checkers later in the future, this could become an issue. The problem is, if we list all the checkers here, the default also gets displayed, so you will see the checkers twice (not to mention the help text line will be *really* long)

This is the help text:

    --checker (-c)        enable the specified checker (default: ["private","deprecation","strong-comparison"]) (multiple values allowed)

What do you think?

nickvergessen replied:
Ah right, I didn't check the output. Seems okay when they output the default value.
Now the question is should we include all in the default (and I think we shouldn't, but private only)?

nickvergessen commented:
Well, that is why I wouldn't add them all by default...
Will have a look whether I can find a way later.

Xenopathic commented:
By default, I think (and @rullzer above mentioned this too) the app developer should have all checks enabled, so they can with minimal effort get the best guidance for their app. Obviously some of the checks are designed to be soft-fails with regards to the app store, so automated checks can enable a subset of the checkers that are designated as hard-fail, e.g. the private checker.

nickvergessen committed:
Only decorate the type when it was matched

nickvergessen commented:
line   80: User::getUser - Method of deprecated class must not be called
    line   89: OC_DB - Static method of private class must not be called

Fixed it

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/1b28a1aa-8777-4b50-ab9b-fccd9da177e8) was created.

nickvergessen commented:
https://github.com/owncloud/core/pull/17046 was merged, so I will add that when jenkins reported back

nickvergessen committed:
PR #17046 deprecated OCP\Util::mb_(sub)str_replace()

MorrisJobke commented:
@owncloud-bot retest this please

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/1face6fe-100c-44cc-8e7f-f9fa0ca086e2) was created.

nickvergessen commented:
Rebased to fix the oracle thing without timeout.

MorrisJobke commented:
@owncloud-bot retest this please

MorrisJobke commented:
@owncloud-bot retest this please

MorrisJobke commented:
Jenkins PR #17720

MorrisJobke commented:
Tested and works nicely :+1:

nickvergessen commented:
@Xenopathic want to review? ;)

Xenopathic commented:
Code looks good, haven't tested it however. :+1:

oparoz commented:
> You can now specify the checkers you want enabled, with --checker=... or -c.... So to only enable private class and strong comparison checks, use ./occ app:check-code -cprivate -cstrong-comparison. By default, all are enabled, so we need to tell Jenkins to only use the private checker at the moment.

Documentation/OP needs to be updated. It still has the old switch, which doesn't work and no explanation on the other valid options

nickvergessen commented:
@oparoz posted in the docs repo: https://github.com/owncloud/documentation/issues/1736

oparoz commented:
Thanks. Added a link to the doc from the list of features.
